### ISSUE TITLE
Consider an OSI-approved license?

### ISSUE THREAD

**Issue by ryan-roemer** [RIAEvangelist/node-ipc#133]
Fantastic library!

While the DBAD license is hilarious, it's sufficiently ambiguous + has some arguable restrictions that means we can't use it with any of our client projects that have a legal department that cares.

Would you be open to considering a more boring-but-widespread [OSI-approved license](https://opensource.org/licenses/category) like MIT or BSD?

Thanks for the potential consideration!

**Comment by jrylan** [comment]
@RIAEvangelist do you see any possibility of this happening? Maybe making it part of a new major release, such as 10.0.0?

I agree 100% with the spirit of the DBAD license, unfortunately legal experts at companies are often pretty ridged about saying "we only accept BSD, ISC and MIT licensed software". 

I've looked around for alternatives and came to the conclusion that `node-ipc` is by far the best cross platform IPC solution for Node.js, so it's unfortunate that some developers can't use it because of the license.

**Comment by dantman** [comment]
I find this fairly annoying as well. It's used in parallel-webpack which I'd like to make use of.

I've got a license checking tool to ensure that I don't accidentally include proprietary or GPL code. It was tripped up by this package because the `license` field is invalid. The `license` field is supposed to be a valid spdx license identifier (or "UNLICENSED" or "SEE LICENSE IN ..." for proprietary code). And "dbad" is not part of the spdx list of licenses.

**Comment by RIAEvangelist** [comment]
Guys the licence is pretty clear, buy me a pint if you or your company is
successful, or forever know deep down inside that you are a dick.  It's
black and white. Enjoy the code wether you are successful or not and buy me
a pint or don't, that is upto you the licence is clear what you are if you
don't and you feel like you should ;)

**Comment by ryan-roemer** [comment]
@RIAEvangelist -- Yep, I don't think any of us are saying you aren't getting your point across. 😄 If I used this package in one of my personal projects, I would definitely buy you a beer or 10.

You have full say on how you want to license this project and allow its use, I just wish that I _could_ use this package for some of my client projects at work (that run software @dantman alludes to, or have mandatory OSS license checklists that licenses have to fall under). But, you call the shots 🍸 so totally understand if keeping with the current license is how you want to roll!

**Comment by jrylan** [comment]
@RIAEvangelist Agreeing with @ryan-roemer here -- we all get it the point and in the context of personal project, or even start ups in which we were the founders there is no issue with the license.

The problem is with corporate employers and clients having a strict set of acceptable licenses. I consult with dozens of companies and the routine is roughly the same: print out the licenses for all the dependencies and subdependencies in `node_modules` and if a license fails their acceptance criteria, you have to remove it, it's not even a debate and you don't get to pushback. 

These companies are in "cover their butts" mode and don't budge in order to avoid issues such as GPL compliance and all that. They are pre-emptively terrified of getting sued over licenses and only allow the few that lawyers have told them are safe, such as MIT, BSD, ISC, etc...I know it's ridiculous, but that's the reality of the situation.

**Comment by dantman** [comment]
Not just companies. This kind of custom licensing is frowned on in the open source community. If the license isn't even listed in spdx (which lists a lot of weird licenses, including things like the old incompatible 4-clause BSD) it's pretty surely not an accepted license. And the wording may even exclude it from most community's definitions of open source. Remember when [a "do not be evil" clause was put Crockford's JSON license](https://blog.sonatype.com/2012/03/use-json-well-youd-better-not-be-evil/).

This license would likely get libraries using it on Apache's Category X list and blacklisted from Debian packages. The terms likely also make it incompatible with any project that is GPL licensed.

**Comment by jrylan** [comment]
Very, very good point there @dantman. This issue echoes the Apache Category X listing of React prior to Facebook relicensing as MIT.

I have a rather large MIT licensed project that's in private alpha testing and I'm currently using `node-ipc `as part of a long running helper client that watches the file system and reports contextual updates across all the connected listeners (Node, VS Code plugin, etc...). `node-ipc` does the job beautifully, but it's currently a lingering task for me to replace the dependency before we officially launch publicly.

It's unfortunate, I'd rather not... but I can't take this risk of anything impact the adoption rate of my project.

**Comment by dantman** [comment]
> Very, very good point there @dantman. This issue echoes the Apache Category X listing of React prior to Facebook relicensing as MIT.

Which is another shame, since React and especially React Native should not have been relicensed to plain MIT. It should have been Apache 2.0, a Apache 2.0/GPL-0.2-or-later dual-license, EPL, or anything else that fits. Just not a MIT license with absolutely no patent protection whatsoever. But that's a separate discussion.

**Comment by RIAEvangelist** [comment]
With the current license, node-ipc is being used in FDA regulated projects
including medical implant software. The licence has not caused any issue
for these regulated fields.

Could you guys site the portion of the current licence which you think
could result in a legal issue for you? If there is in fact a possible legal
issue I will then open up to suggestions. As it stands, I do not believe
the licence gives me any claims on anything other than calling someone a
dick if they do not purchase me a pint.

I reserve the right to be wrong, and give you the opportunity to educate me
on the actual LEGAL reason for concern. If, however, the reason for concern
is *only* that it is not one of the more popular licences alone, than that
is a good reason to evaluate your current stance. I am not a fan of "thats
how we have always done it". Hence the fact that I am open to hearing you
guys out.

**Comment by dantman** [comment]
https://en.wikipedia.org/wiki/License_proliferation
https://opensource.org/node/365
https://opensource.org/proliferation

**Comment by ryan-roemer** [comment]
There definitely is one level of just "predictability" and "it's already approved" that makes things easier. But you don't have to be boring just because everyone else is.

... so I'd look at the next issue of "restrictions". A license gives up certain copyrights within certain bounds. For MIT, BSD, etc. there's not much in the way of restrictions beyond preserving some notices, etc. By contrast, for GPL, there are restrictions like situationally having to release the source of something using a GPL-ified work.

## Analysis

So for DBAD, the actual restrictions I could potentially see are:

- Section 1: you can't be a dick. Or else, presumably, you lose the license to the work and you are infringing copyright. Because 1a-c enumerate only _some_ ("[b]eing a dick includes - but is not limited to") there is a lot of open, other places for someone to be considered "a dick" and thus losing the license.
    - This then raises a bigger issue: who is evaluating "dick-ish-ness"? Presumably it's the license holder. So jokes aside, this section could be construed as @RIAEvangelist has a subjective determination of "dickishness" by any `node-ipc` user at their sole discretion in ways that aren't already known.
    - And any one person's idea of what a "dick" may be vastly different than anothers (@RIAEvangelist's in particular), or what a court / jury may eventually have to decide (which is the ultimate target audience of any copyright license).

- Section 2: "If you become rich ... share the love. ... not buy the original work's creator(s) a pint."
    - This potentially means any amount of profit derived in some portion from `node-ipc` then creates an obligation to compensate @RIAEvangelist in some fashion, at minimum buying a beer.
    - While this is undeniably reasonable, it also means that a company's legal department _should_ have an action plan to both (1) evaluate if any profit is made that can be attributed in part to `node-ipc` and (2) then coordinate at least a beer for @RIAEvangelist
    - ... or else you are infringing copyright law because the license is lost.
    - This one certainly isn't an onerous obligation, but it has a logistical cost to it to ensure compliance. And it's not something the _developers_ (your target audience) can necessarily ensure (as opposed to like "preserve license header comments" which they can.)

- Section 3: Probably OK. Offhand, I'm not seeing anything that big in the way of creating an actual restriction (?)

Not sure exactly if the above is right, or what that means to you in terms of what a refactor, if any, would look like.

## Maybe (?)

That is why it's so easy to go with an OSI-approved license. Part of the things the OSI folks do is certify that restrictions/limitations within a license end up in a place that ["they allow software to be freely used, modified, and shared."](https://opensource.org/licenses) 

But, for this license, I'm not exactly sure... It could be something like having the operative language be the allowance "Do whatever you like with the original work", but without the **actual** restrictions.

So, maybe something like:

1. Do whatever you like with the original work, just don't be a dick. You don't lose any rights to this license if you are a dick, but you will know that you suck and are a dick.
2. If you become rich through this work, you should share the love. You don't have to, but you will know that you suck and are a dick.
3. ( _Probably can leave unmodified_ )

Anyways, in conclusion, this is quite possibly the strangest GitHub comment I have ever written. I hope it helps. 😄

**Comment by RIAEvangelist** [comment]
Where can I send you a pint? Your detailed responses was both hilarious and
informative. I will modify the license as you have suggested, and submit
the changes to the original work as well sighting this thread.

**Comment by dantman** [comment]
Fixing questionable legal wordings in the license is good. But as far as Open Source goes having developers make their own custom license is still harmful to the ecosystem, even if there were made by an actual lawyer. Developer A trying to use your library in their project is not qualified to determine whether your license is compatible with their license and acceptable to all the people they want to use their code.

To be honest, the only reason I don't care about the resolution to this anymore is because the alpha for `parallel-webpack`'s next major version has already dropped the dependency that uses `node-ipc` and I have no plans to use `node-ipc` in a future open source project.

Anyways, if you are going to keep the DBAD license please at least fix your package.json. `DBAD` isn't registered with spdx meaning that `"license": "DBAD"` is invalid according to the [npm documentation](https://docs.npmjs.com/files/package.json#license) so change it to `"SEE LICENSE IN licence.md"`.

**Comment by ZeeCoder** [comment]
"Being a dick includes - but is not limited to - the following instances:

1a. Outright copyright infringement - Don't just copy this and change the name."

So if I were to copy the lib, change the name BUT also change the license (say, to MIT) and the readme, saying that the only reason I'm doing this is to allow myself to use it for client projects (so giving full credit to the author), and also were to buy you a pint, according to this license that seems fine to me? 🤔

**Comment by oldmud0** [comment]
Why not dual license it? :laughing:

**Comment by ZeeCoder** [comment]
oh so dual licensing means either or? not that both applies? 😲

**Comment by hyandell** [comment]
Noting that the license is now on the Apache Category X list.

**Comment by oldmud0** [comment]
I also wish to note that I do not use this library any more - there are also style issues I have experienced working with this library. I am using websockets for IPC instead, which is good enough for my small project (I'd recommend a message broker/message queue like Redis or ZeroMQ for large projects though), but that's a whole different topic altogether.

**Comment by jefftherobot** [comment]
I'd like to add my comment and get an spdx approved license here. This package is used by @vue/cli and I can't get it through my corporate policy license checks. Where did this land?

**Comment by RIAEvangelist** [comment]
I will update to Apache 2.0 in v10.x which is underway no on the v10.x branch

Reopening until released.

**Comment by RIAEvangelist** [comment]
Switched to MIT

**Comment by joshuaalayon** [comment]
Be aware - this package was found to contain malware that detects its users machines purported geographical location and replaces the data from the machine if it does not like the location it detects

If you are still using this package - you may do better by locking it to an older version or, preferably, replacing it altogether along with all software maintained by the repos maintainer(s)

https://www.bleepingcomputer.com/news/security/big-sabotage-famous-npm-package-deletes-files-to-protest-ukraine-war/amp/
